.github/workflows/r-ci.yml

```diff
@@ -52,6 +53,63 @@ jobs:
       - name: Bootstrap
         run: cd apis/r && tools/r-ci.sh bootstrap
 
+      - name: Set additional repositories (macOS)
+        if: ${{ matrix.os != 'ubuntu-latest' }}
+        run: echo 'options(repos = c("https://tiledb-inc.r-universe.dev", getOption("repos")))' | tee -a ~/.Rprofile
+
+      - name: Set additional repositories (Linux)
+        if: ${{ matrix.os == 'ubuntu-latest' }}
+        run: |
+          rversion <- paste(strsplit(as.character(getRversion()), split = '\\.')[[1L]][1:2], collapse = '.')
+          codename <-  system('. /etc/os-release; echo ${VERSION_CODENAME}', intern = TRUE)
+          repo <- "https://tiledb-inc.r-universe.dev"
+          (opt <- sprintf('options(repos = c("%s/bin/linux/%s/%s", "%s", getOption("repos")))', repo, codename, rversion, repo))
+          cat(opt, "\n", file = "~/.Rprofile", append = TRUE)
+        shell: Rscript {0}
+
+      - name: Install tiledb-r
+        run: |
+          options(bspm.version.check = TRUE)
+          # Find tiledb-r constraints
+          deps <- read.dcf('apis/r/DESCRIPTION', fields = 'Imports')[1, ]
+          deps <- trimws(strsplit(gsub(pattern = ',', replacement = '\n', x = deps), split = '\n')[[1L]])
+          deps <- Filter(nzchar, deps)
+          deps <- Filter(function(x) grepl('^tiledb', x), deps)
+          deps <- gsub(pattern = '[[:space:]]', replacement = '', x = deps)
+          deps <- gsub(pattern = '.*\\(|\\)', replacement = '', x = deps)
+          const <- data.frame(comp = gsub(pattern = '[[:digit:]\\.]*', replacement = '', x = deps))
+          const$vers <- gsub(pattern = paste(const$comp, collapse = '|'), replacement = '', x = deps)
+          const
+          # Find correct version of tiledb-r
+          db <- utils::available.packages(filters = c("R_version", "OS_type", "subarch"))
+          idx <- which(rownames(db) == 'tiledb')
+          valid <- vapply(
+            idx,
+            FUN = function(i) {
+              v <- db[i, 'Version']
+              res <- vector(mode = 'logical', length = nrow(const))
+              for (i in seq_along(res)) {
+                res[i] <- do.call(const$comp[i], args = list(v, const$vers[i]))
+              }
+              return(all(res))
+            },
+            FUN.VALUE = logical(1L)
+          )
+          if (!any(valid)) {
+            stop("No valid versions of tiledb-r found")
+          }
+          db <- db[-idx[!valid], , drop = FALSE]
+          # Install upstream deps
+          ups <- tools::package_dependencies("tiledb", db = db, recursive = TRUE)$tiledb
+          utils::install.packages(intersect(ups, rownames(db)))
+          # Install correct version of tiledb-r
+          ctb <- contrib.url(getOption("repos"))
+          names(ctb) <- getOption("repos")
+          dbr <- db[idx[valid], 'Repository']
+          repos <- names(ctb[ctb %in% dbr])
+          utils::install.packages("tiledb", repos = repos)
```

Comment:
This is IMHO hard to read (inside YAML) and impossible to test.  Can we make it standalone script?

Also the install step here (before dependencies) may pull in a few other things at this step.

Reply:
I've created three install steps
1. install upstream hard dependencies of tiledb-r https://github.com/single-cell-data/TileDB-SOMA/pull/2540/files#diff-5331055c5808c739bfd30c63013fd35ce1e28e3edef1df4476550617d393467cR102-R104
2. install tiledb-r from the specified repository https://github.com/single-cell-data/TileDB-SOMA/pull/2540/files#diff-5331055c5808c739bfd30c63013fd35ce1e28e3edef1df4476550617d393467cR110
3. install dependencies of tiledbsoma-r with an explicit `upgrade = "never"` to prevent tiledb-r and its dependencies from being upgraded https://github.com/single-cell-data/TileDB-SOMA/pull/2540/files#diff-5331055c5808c739bfd30c63013fd35ce1e28e3edef1df4476550617d393467cR143

I can move the "Install tiledb-r" step into a standalone script

Reply:
Moved the installation steps into `apis/r/tools/install-tiledb-r.R` in [`88bafc4`](https://github.com/single-cell-data/TileDB-SOMA/pull/2540/commits/88bafc4213d624be8a2e6ff0f3cf3d2d253a77cc)